### PR TITLE
Enable -Wshadow by default

### PR DESCRIPTION
**Pull request by paveljanik**
Test and enabled `-Wshadow` by default, add developer note about this change (see #8105).

As my native language is not English, I'm open to suggestions.

[Edits from maintainers allowed for easier cooperation.]

### CODE REVIEW

File: doc/developer-notes.md
Comment (laanwj) [inline doc/developer-notes.md]: Please mention the rationale why it is enabled, why we went through this in the first place: which is to avoid bugs like #8102

Comment (paveljanik) [inline doc/developer-notes.md]: Done in the squashme commit.

**Comment by MarcoFalke** [comment]
Concept ACK

**Review by dcousens — APPROVED** [review]

**Comment by laanwj** [comment]
utACK.
At this moment, though, this still generates in the process of the build:
```
$ grep Wshadow /tmp/log|wc
   3406   30643  458032
```
I don't think this number needs to be 0 before merging this, but it does needs to be a manageable number to avoid 'shadowing' other warnings and errors, as well as increasing the Travis output by a thousandfold.

**Comment by paveljanik** [comment]
Definitely. This has to wait.

There are other PRs open, serialization changes has to be merged, and I still have to PR txmempool, main, wallet, torcontroller and tests changes.

**Comment by laanwj** [comment]
I'd say most urgent is to remove shadowing in header files, especially the often-included ones. Those warnings come back for every compiled file.

**Comment by laanwj** [comment]
The univalue upstream merge (#8807) brought the total number of warnings here down to 3197.

**Comment by paveljanik** [comment]
@laanwj "here" means with gcc, other PRs (see the end of #8105) included? I'm now targeting gcc to be shadow free...

**Comment by laanwj** [comment]
Current master w/ clang 3.9.0, full build (including GUI). No extra PRs applied.

**Comment by laanwj** [comment]
Current master (after merging #8655) brought the number of Wshadow warnings in a full build (with clang) down to 3167

**Comment by paveljanik** [comment]
@laanwj Can you upload the build log after `grep "warning: declaration shadows"` somewhere for me?

**Comment by laanwj** [comment]
@paveljanik Sure: https://dev.visucore.com/bitcoin/tmp/log.gz

I think what is left is nearly all serialization related, certainly the ones that appear many times.

**Comment by paveljanik** [comment]
Great - this means there are 42 unique instances of this warning. Getting closer to zero :-)

**Comment by fanquake** [comment]
Concept ACK. Optimistically tagging this for 0.14.0

**Comment by fanquake** [comment]
Running master + #9039 I only see one -Wshadow warning on OS X, which is fixed by #8981.

**Comment by paveljanik** [comment]
@fanquake Thanks for testing, Michael!

**Comment by MarcoFalke** [comment]
Since the two pulls got merged, anything holding this back besides maybe #8808?

**Comment by laanwj** [comment]
Yes, thanks for testing @fanquake. 
I can confirm the same with LLVM 4.0 (git master, empty ccache).

As #9039 is mergedk this is ready for merge after the squashme is squashed.

**Comment by fanquake** [comment]
Just retested with [master](https://github.com/bitcoin/bitcoin/commit/e81df49644c21f835e25028ab4643aa9bf5ae8da) + this PR. Can confirm that there are no -Wshadow warnings.

**Comment by paveljanik** [comment]
Thanks everyone for testing!

**Comment by laanwj** [comment]
Going to revert this, even though I tested this with two compilers before merging, people on  IRC are apperently noticing floods of Wshadow warnings.

**Pull request merged**
